cluster/replication/consumer.go

```diff
@@ -165,13 +165,35 @@ func (c *CopyOpConsumer) Consume(ctx context.Context, in <-chan ShardReplication
 
 				// Check if the operation is already in progress
 				if c.ongoingOps.LoadOrStore(op.Op.ID) {
-					// Avoid scheduling unnecessary work or incorrectly counting metrics
-					// for operations that are already in progress or completed.
-					c.logger.WithFields(logrus.Fields{"consumer": c, "op": op}).Debug("replication op skipped as already running")
-					// Need to release the token to let other consumers process queued replication operations.
-					<-c.tokens
-					c.engineOpCallbacks.OnOpSkipped(c.nodeId)
-					continue
+					if !op.Status.ShouldCancelOrDelete() {
+						// Avoid scheduling unnecessary work or incorrectly counting metrics
+						// for operations that are already in progress or completed.
+						c.logger.WithFields(logrus.Fields{"consumer": c, "op": op}).Debug("replication op skipped as already running")
+						// Need to release the token to let other consumers process queued replication operations.
+						<-c.tokens
+						c.engineOpCallbacks.OnOpSkipped(c.nodeId)
+						continue
+					}
+					// This means we're cancelling/deleting the operation, so we need to call the op's cancel
+					cancel, ok := c.ongoingOps.LoadCancel(op.Op.ID)
+					if ok {
+						c.logger.WithFields(logrus.Fields{"consumer": c, "op": op}).Debug("cancelling the replication op")
+						// If there's something to cancel then cancel it
+						cancel()
+						// Then signify that the operation was cancelled so that the state can be updated in goroutine process
+						// once dispatchReplicationOp exits
+						c.ongoingOps.StoreCancelled(op.Op.ID)
+					} else {
+						// Otherwise, update the state to cancelled immediately now since we don't have a cancel function
+						// meaning that any in-flight operation has already completed
```

Comment:
is this handling the case where the in-flight operation completed successfully or was cancelled or deleted or something else? i ask because if the operation was successful or deleted, i don't understand why we would update the status to cancelled in that case, but i assume i'm misunderstanding why we would not have a cancel function

Reply:
So, this is handling the case where the in-flight process exits, returning the token and calling `c.ongoingOps.Remove(op.Op.ID)` inside the `defer` on L208, while another goroutine is attempting to cancel said process. In such a case, there will be no `cancel` function but the `shouldCancel` flag in the status of the op will be true. As such, it is the consumer's responsibility to signal that the state is cancelled.

If, on the other hand, the in-flight process hasn't exited yet then the call to `cancel()` will cause `stateFuncHandler` to exit early with an error that is handled appropriately in the `processStateAndTransition` function. I refactored the logic to make this simpler and clearer based on your comment 😁